File: easysql_api/models/query.py
from __future__ import annotations

from typing import Any

from pydantic import BaseModel, Field

from easysql_api.domain.value_objects.query_status import QueryStatus


class QueryRequest(BaseModel):
    question: str = Field(..., min_length=1, max_length=2000)
    db_name: str | None = Field(default=None)
    session_id: str | None = Field(default=None)
    stream: bool = Field(default=False)


class ContinueRequest(BaseModel):
    answer: str = Field(..., min_length=1, max_length=1000)
    stream: bool = Field(default=False)
    thread_id: str | None = Field(default=None, description="Thread ID for branch isolation")


class MessageRequest(BaseModel):
    question: str = Field(..., min_length=1, max_length=2000)
    stream: bool = Field(default=False)
    parent_message_id: str | None = Field(default=None, description="Parent message ID")
    thread_id: str | None = Field(default=None, description="Thread ID for branch isolation")


class BranchRequest(BaseModel):
    from_message_id: str = Field(..., description="Message ID to branch from")
    question: str = Field(..., min_length=1, max_length=2000)
    stream: bool = Field(default=False)
    thread_id: str | None = Field(default=None, description="Parent thread ID for branching")


class ForkSessionRequest(BaseModel):
    from_message_id: str | None = Field(default=None, description="Source assistant message ID")
    thread_id: str | None = Field(default=None, description="Source thread ID for branching")
    turn_ids: list[str] = Field(default_factory=list, description="Ordered turn IDs to clone")


class ClarificationInfo(BaseModel):
    questions: list[str]


class QueryResponse(BaseModel):
    session_id: str
    status: QueryStatus
    sql: str | None = None
    validation_passed: bool | None = None
    validation_error: str | None = None
    clarification: ClarificationInfo | None = None
    error: str | None = None
    stats: dict[str, Any] | None = None
    message_id: str | None = None
    parent_message_id: str | None = None
    thread_id: str | None = None


class StreamEvent(BaseModel):
    event: str
    data: dict[str, Any]

File: easysql_web/src/types/query.ts
import type { VizPlan } from './chart';

export interface QueryRequest {
  question: string;
  db_name?: string;
  session_id?: string;
  stream?: boolean;
}

export interface ContinueRequest {
  answer: string;
  stream?: boolean;
  thread_id?: string;
}

export interface ClarificationInfo {
  questions: string[];
}

export type QueryStatus = 
  | 'pending'
  | 'processing'
  | 'awaiting_clarify'
  | 'completed'
  | 'failed';

export interface QueryResponse {
  session_id: string;
  status: QueryStatus;
  sql?: string;
  validation_passed?: boolean;
  validation_error?: string;
  clarification?: ClarificationInfo;
  error?: string;
  stats?: Record<string, unknown>;
  message_id?: string;
  parent_message_id?: string;
  thread_id?: string;
  turn_id?: string;
}

export interface StreamEvent {
  event: 'start' | 'state_update' | 'agent_progress' | 'complete' | 'error';
  data: {
    node?: string;
    session_id?: string;
    thread_id?: string;
    message_id?: string;
    parent_message_id?: string;
    turn_id?: string;
    generated_sql?: string;
    status?: QueryStatus;
    validation_passed?: boolean;
    validation_result?: {
      valid: boolean;
      details?: string;
      error?: string;
    };
    clarification_questions?: string[];
    clarification?: {
      questions: string[];
    };
    retrieval_summary?: {
      tables_count: number;
      tables: string[];
    };
    context_summary?: {
      total_tokens: number;
      has_system_prompt: boolean;
      has_user_prompt: boolean;
    };
    chart_plan?: VizPlan;
    chart_reasoning?: string;
    sql?: string;
    error?: string;
    type?: 'tool_start' | 'tool_end' | 'thinking' | 'token' | 'thought_complete';
    iteration?: number;
    action?: 'tool_start' | 'tool_end' | 'thinking';
    tool?: string;
    success?: boolean;
    input_preview?: string;
    output_preview?: string;
    content?: string;
  };
}

export interface MessageRequest {
  question: string;
  stream?: boolean;
  parent_message_id?: string;
  thread_id?: string;
}

export interface BranchRequest {
  from_message_id: string;
  question: string;
  stream?: boolean;
  thread_id?: string;
}

export interface ForkSessionRequest {
  from_message_id?: string;
  thread_id?: string;
  turn_ids?: string[];
}

export interface ForkSessionResponse extends QueryResponse {
  source_session_id: string;
  cloned_turn_ids: string[];
}

export interface MessageResponse {
  session_id: string;
  message_id?: string;
  parent_message_id?: string;
  thread_id?: string;
  status: QueryStatus;
  sql?: string;
  validation_passed?: boolean;
  error?: string;
}
